File: leetcode/src/python/141.linked-list-cycle.py
#
# @lc app=leetcode id=141 lang=python3
#
# [141] Linked List Cycle
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        # 龟兔赛跑
        fast = slow = head
        while slow and fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow is fast:
                return True
        
        return False
        
# @lc code=end

File: leetcode/src/java/141.linked-list-cycle.java
/*
 * @lc app=leetcode id=141 lang=java
 *
 * [141] Linked List Cycle
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    // 环检测
    // 记住龟兔赛跑
    public boolean hasCycle(ListNode head) {
        ListNode lower = head;
        ListNode fast = head;

        while (lower != null && fast != null && fast.next != null) {
            lower = lower.next;
            fast = fast.next.next;
            if (lower == fast) {
                return true;
            }
        }
        return false;
    }
}
// @lc code=end

